server/store/datastore/migration/copy.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"xorm.io/xorm"
+	"xorm.io/xorm/schemas"
+
+	"go.woodpecker-ci.org/woodpecker/v2/server/model"
+)
+
+func Copy(ctx context.Context, src, dest *xorm.Engine) error {
+	// first check if the new database already has existing data
+	for _, bean := range allBeans {
+		exist, err := dest.IsTableExist(bean)
+		if err != nil {
+			return err
+		} else if exist {
+			return fmt.Errorf("existing table '%s' in import destination detected", dest.TableName(bean))
+		}
+	}
+
+	// next we make sure the all required migrations are executed
+	if err := Migrate(ctx, src, true); err != nil {
+		return fmt.Errorf("migrate source database failed: %w", err)
+	}
+
+	// init schema in destination
+	if err := initSchemaOnly(dest); err != nil {
+		return err
+	}
+
+	// copy data
+	// IMPORTANT: if you add something here, also add it to migration.go allBeans slice
```

Comment:
sadly this wont work, i tried hard to ose the `any` type ... or reflections.

but i could not create a list of items based on `any` ... if i dont batch and itterate over each item by one it might work ... but that is sure inefficient!

Reply:
xorm per se can not do anything ... it would have to add an Database-Migration-Feature.
that will work

Reply:
https://gitea.com/xorm/xorm/issues/2478

Reply:
the problem is that then you have to validate each item in the list ... that slows down the whole orm!